src/main/groovy/lang/IntRange.java

```diff
@@ -385,26 +385,23 @@ public void step(int step, Closure closure) {
             return; // from == to and step == 0, nothing to do, so return
         }
 
-        if (isReverse()) {
-            step = -step;
+        final long from = getFrom().longValue(), to = getTo().longValue();
+
+        final boolean isAscending;
+        if (step < 0) {
+            step *= -1;
+            isAscending = isReverse();
+        } else {
+            isAscending = !isReverse();
         }
-        if (step > 0) {
-            int value = getFrom();
-            while (value <= getTo()) {
+
+        if (isAscending) {
+            for (long value = from; value <= to; value += step) {
                 closure.call(value);
-                if (((long) value + step) >= Integer.MAX_VALUE) {
-                    break;
-                }
-                value = value + step;
             }
         } else {
-            int value = getTo();
-            while (value >= getFrom()) {
+            for (long value = to; value >= from; value -= step) {
```

Comment:
And of course the most simple fix is to cast the long to int here, just before calling the closure. since the long is only for overflow, it should be perfectly valid to do so